common/djangoapps/student/tests/tests.py

```diff
@@ -930,3 +935,230 @@ def test_roundtrip_with_unicode_course_id(self):
         real_user = user_by_anonymous_id(anonymous_id)
         self.assertEqual(self.user, real_user)
         self.assertEqual(anonymous_id, anonymous_id_for_user(self.user, course2.id, save=False))
+
+
+@unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')
+@ddt.ddt
+class DashboardTestXSeriesPrograms(ModuleStoreTestCase, ProgramsApiConfigMixin):
+    """
+    Tests for dashboard for xseries program courses. Enroll student into
+    programs and then try different combinations to see xseries upsell
+    messages are appearing.
+    """
+    def setUp(self):
+        super(DashboardTestXSeriesPrograms, self).setUp()
+
+        self.user = UserFactory.create(username="jack", email="jack@fake.edx.org", password='test')
+        self.course_1 = CourseFactory.create()
+        self.course_2 = CourseFactory.create()
+        self.course_3 = CourseFactory.create()
+        self.program_name = 'Testing Program'
+        self.category = 'xseries'
+
+        CourseModeFactory.create(
+            course_id=self.course_1.id,
+            mode_slug='verified',
+            mode_display_name='Verified',
+            expiration_datetime=datetime.now(pytz.UTC) + timedelta(days=1)
+        )
+        self.client = Client()
+        cache.clear()
+
+    def _create_program_data(self, data):
+        """Dry method to create testing programs data."""
+        programs = {}
+        for course, program_status in data:
+            programs[unicode(course)] = {
+                'category': self.category,
+                'organization': {'display_name': 'Test Organization 1', 'key': 'edX'},
+                'marketing_slug': 'fake-marketing-slug-xseries-1',
+                'status': program_status,
+                'course_codes': [
+                    {
+                        'display_name': 'Demo XSeries Program 1',
+                        'key': 'TEST_A',
+                        'run_modes': [{'sku': '', 'mode_slug': 'ABC', 'course_key': unicode(course)}]
+                    },
+                    {
+                        'display_name': 'Demo XSeries Program 1',
+                        'key': 'TEST_A',
+                        'run_modes': [{'sku': '', 'mode_slug': 'ABC', 'course_key': unicode(course)}]
+                    },
+                    {
+                        'display_name': 'Demo XSeries Program 1',
+                        'key': 'TEST_A',
+                        'run_modes': [{'sku': '', 'mode_slug': 'ABC', 'course_key': unicode(course)}]
+                    }
```

Comment:
why repeat the dictionary 3x?  we will never see duplicated course codes or run modes in real data.